src/vm/method.cpp

```diff
@@ -4798,6 +4811,214 @@ Precode* MethodDesc::GetOrCreatePrecode()
+    }
+
+    EntryPointSlotsToBackpatch::Backpatch_Locked(slot, slotType, currentEntryPoint);
+}
+
+void MethodDesc::BackpatchEntryPointSlots(PCODE entryPoint, bool isPrestubEntryPoint)
+{
+    WRAPPER_NO_CONTRACT;
+    _ASSERTE(entryPoint != NULL);
+    _ASSERTE(IsEligibleForEntryPointSlotBackpatch());
+    _ASSERTE(isPrestubEntryPoint == (entryPoint == GetPrestubEntryPointToBackpatch()));
+
+    LoaderAllocator *mdLoaderAllocator = GetLoaderAllocator();
+    MethodDescBackpatchInfoTracker *backpatchInfoTracker = mdLoaderAllocator->GetMethodDescBackpatchInfoTracker();
+    MethodDescBackpatchInfoTracker::ConditionalLockHolder lockHolder;
+
+    // Get the entry point to backpatch inside the lock to synchronize with backpatching in MethodDesc::DoBackpatch()
+    if (GetEntryPointToBackpatch_Locked() == entryPoint)
+    {
+        return;
+    }
+
+    if (IsTieredVtableMethod())
+    {
+        // Backpatch the func ptr stub if it was created
+        FuncPtrStubs *funcPtrStubs = mdLoaderAllocator->GetFuncPtrStubsNoCreate();
+        if (funcPtrStubs != nullptr)
+        {
+            Precode *funcPtrPrecode = funcPtrStubs->Lookup(this);
+            if (funcPtrPrecode != nullptr)
+            {
+                if (isPrestubEntryPoint)
+                {
+                    funcPtrPrecode->ResetTargetInterlocked();
+                }
+                else
+                {
+                    funcPtrPrecode->SetTargetInterlocked(entryPoint, FALSE /* fOnlyRedirectFromPrestub */);
+                }
+            }
+        }
+    }
+
+    MethodDescBackpatchInfo *backpatchInfo = backpatchInfoTracker->GetBackpatchInfo_Locked(this);
+    if (backpatchInfo != nullptr)
+    {
+        // Backpatch slots from the same loader allocator
+        backpatchInfo->GetSlotsToBackpatch()->Backpatch_Locked(entryPoint);
+
+        // Backpatch slots from dependent loader allocators
+        backpatchInfo->ForEachDependentLoaderAllocatorWithSlotsToBackpatch_Locked(
+            [&](LoaderAllocator *slotLoaderAllocator)
+        {
+            _ASSERTE(slotLoaderAllocator != nullptr);
+            _ASSERTE(slotLoaderAllocator != mdLoaderAllocator);
+
+            EntryPointSlotsToBackpatch *slotsToBackpatch =
+                slotLoaderAllocator
+                ->GetMethodDescBackpatchInfoTracker()
+                ->GetDependencyMethodDescEntryPointSlotsToBackpatch_Locked(this);
+            if (slotsToBackpatch != nullptr)
+            {
+                slotsToBackpatch->Backpatch_Locked(entryPoint);
+            }
+        });
+    }
+
+    // Set the entry point to backpatch inside the lock to synchronize with backpatching in MethodDesc::DoBackpatch(), and set
+    // it last in case there are exceptions above, as setting the entry point indicates that all recorded slots have been
+    // backpatched
+    SetEntryPointToBackpatch_Locked(entryPoint, isPrestubEntryPoint);
+}
+
+void MethodDesc::SetCodeEntryPoint(PCODE entryPoint)
```

Comment:
I didn't see a good way to fully avoid ambiguity while keeping the names simple, I also think there is already some ambiguity in the names. There are SetMethodEntryPoint(), SetStableEntryPointInterlocked(), SetNativeCodeInterlocked(), SetCodeEntryPoint(), PublishNativeCode().

In my view SetMethod... and SetStable... refer to the initial entry point that will be called. That may be the code entry point, it may be a precode that redirects to the code entry point, it may be a some other stub, etc.

SetCode... refers to the actual code entry point that the initial entry point will eventually redirect to. When the method's entry point is the code entry point, SetCode... would update the method's entry point. When the method's entry point is a precode, SetCode... would update the precode's target. SetCode... also handles backpatching other slots that also point directly to the method's code entry point.

SetNativeCode... seems to be an odd mix. If there is a native code slot, it just stores the code entry point in the slot without updating any actual entry point, but the stored code entry point may be used by callers. The prestub may later update the method's entry point with the stored code entry point. Tiering uses the native code slot as a storage location for tier 0 code, so SetNativeCode... cannot be called for other code versions. If there is no native code slot, SetNativeCode... does something similar to SetCode... but for more specific cases.

Of the lot, I'm actually most satisfied with the name SetCodeEntryPoint(). I think I'll leave it as is for now, maybe something can be done in the future to rename or refactor these functions to decrease ambiguity and simplify/clarify their purpose.

Reply:
It is possible to define the terms a bit differently and less restrictively while still keeping true to the sense of the terms.

EntryPoint could be defined as anything that can be called by a caller using an appropriate ABI for that EntryPoint and MethodDesc to execute the method correctly. For the purposes of discussion at least, an indirection cell could also be considered an EntryPoint if it contains a value that is an EntryPoint, it is just an indirect EntryPoint. Under that definition the results of some cases above may be different:
- The start of jitted or prejitted code for a method typically is an EntryPoint
- A Precode is an EntryPoint
- The target of a Precode may be an EntryPoint
- A stub may or may not be an EntryPoint depending on its ABI requirements. For example, a Prestub is not an EntryPoint, but a CallCountingStub for tiering could be an EntryPoint.
- For code versionable with a JumpStamp, the start of jitted or prejitted code for any version is an EntryPoint. Though it is preferable that the first version is used in multi-callable cases to maintain versionability, for a single call it may be correct to call a different version directly.

In SetCodeEntryPoint(), what I meant by Code is the start of any code that immediately begins executing the method's code.
- The start of jitted or prejitted code for the method is a CodeEntryPoint
- A Precode may or may not be a CodeEntryPoint. If the target of the Precode is a CodeEntryPoint, then the Precode is not a CodeEntryPoint. If on the other hand, the ABI requires calls to go through the Precode, then the Precode is a CodeEntryPoint.
- The target of a Precode may be a CodeEntryPoint
- An entry point into the interpreter that starts executing method's IL is a CodeEntryPoint
- A CallCountingStub could be considered a CodeEntryPoint if it is considered that it is just an implementation detail that a stub is used. Call counting may be considered part of the method's code as a whole, just with that portion separated into a stub.

NativeCode could be used for cases that specifically refer to compiled code. A MethodDesc could be agnostic about the type of code that a code slot contains. For instance, it may be useful to store an interpreter entry point in the NativeCodeSlot but it currently would be a violation of the name.

I'll go ahead and leave it as is for now. Once we figure out what we want the names to be, the names can be cleaned up in a separate PR.

Reply:
Sounds good to me. Thanks Kount!